src/runtime_src/core/pcie/emulation/hw_em/alveo_shim/plugin/xdp/pl_deadlock.cpp

```diff
@@ -0,0 +1,65 @@
+/**
+ * Copyright (C) 2021 Xilinx, Inc
+ * Copyright (C) 2022 Advanced Micro Devices, Inc. - All rights reserved
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"). You may
+ * not use this file except in compliance with the License. A copy of the
+ * License is located at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include "pl_deadlock.h"
+#include "core/common/module_loader.h"
+#include "core/common/dlfcn.h"
+#include <iostream>
+
+namespace xdp {
+namespace pl_deadlock {
+
+  void load()
+  {
+    static xrt_core::module_loader xdp_pl_deadlock_loader("xdp_pl_deadlock_plugin",
+                register_callbacks,
+                warning_callbacks);
+  }
+  std::function<void (void*)> update_device_cb;
+  std::function<void (void*)> flush_device_cb;
```

Comment:
Why not using just a pointer, since you are not using any feature of `std::function`?

Reply:
std::function is subjectively more reader friendly.

Reply:
If you want to be subjectively reader friendly, add some comments to the code. :-)
`std::function` is a complex object to do some type erasure with some dynamic memory allocation.
Yes, it can also emulate a pointer. Of course here this is about hardware emulation, so everything is already slow, so why not making it slower? ;-)
At least if you use a `std::function`, comparing to `nullptr` instead of just using `if (update_device_cb)` is kind of confusing, even accepted by the C++ standard.

Reply:
Any recommendation has to be read in a given context.
It would be interesting to share here the links to the pieces of advice made about replacing plain pointers to functions by just `std::function` so we can look at it.